sdk/keyvault/azure-security-keyvault-secrets/test/ut/secret_client_base_test.hpp

```diff
@@ -0,0 +1,238 @@
+ *
+ */
+#include <gtest/gtest.h>
+
+#include <azure/core/test/test_base.hpp>
+#include <azure/identity/client_secret_credential.hpp>
+#include <azure/keyvault/keyvault_secrets.hpp>
+namespace Azure { namespace Security { namespace KeyVault { namespace Secrets { namespace _test {
+  /**
+   * @brief Client Secret Credential authenticates with the Azure services using a Tenant ID, Client
+   * ID and a client secret.
+   *
+   */
+  class TestClientSecretCredential final : public Core::Credentials::TokenCredential {
+  public:
+    Core::Credentials::AccessToken GetToken(
+        Core::Credentials::TokenRequestContext const& tokenRequestContext,
+        Core::Context const& context) const override
+    {
+      Core::Credentials::AccessToken accessToken;
+      accessToken.Token = "magicToken";
+      accessToken.ExpiresOn = DateTime::max();
+
+      if (context.IsCancelled() || tokenRequestContext.Scopes.size() == 0)
+      {
+        accessToken.ExpiresOn = DateTime::min();
+      }
+
+      return accessToken;
+    }
+  };
+  class KeyVaultSecretClientTest : public Azure::Core::Test::TestBase,
+                                   public ::testing::WithParamInterface<int> {
+  private:
+    std::unique_ptr<Azure::Security::KeyVault::Secrets::SecretClient> m_client;
+    std::string GetEnv(const std::string& name, std::string const& defaultValue = std::string())
+    {
+#if defined(_MSC_VER)
+#pragma warning(push)
+#pragma warning(disable : 4996)
+      const char* ret = std::getenv(name.data());
+#pragma warning(pop)
+#else
+      const char* ret = std::getenv(name.data());
+#endif
+
+      if (!ret)
+      {
+        if (defaultValue.size() > 0)
+        {
+          return defaultValue;
+        }
+
+        throw std::runtime_error(
+            name + " is required to run the tests but not set as an environment variable.");
+      }
+
+      return std::string(ret);
+    }
+
+  protected:
+    int m_testPollingTimeOutMinutes = 20;
+    std::chrono::minutes m_testPollingIntervalMinutes = std::chrono::minutes(1);
+
+    std::shared_ptr<Azure::Identity::ClientSecretCredential> m_credential;
+    std::shared_ptr<TestClientSecretCredential> m_testCredential;
+    std::string m_keyVaultUrl;
+    std::string m_keyVaultHsmUrl;
+
+    Azure::Security::KeyVault::Secrets::SecretClient const& GetClientForTest(
+        std::string const& testName,
+        Azure::Core::Test::TestMode const& testMode)
+    {
+      std::string testModeValue = "LIVE";
+
+      switch (testMode)
+      {
+        case Azure::Core::Test::TestMode::RECORD: {
+          testModeValue = "RECORD";
+        }
+        break;
+        case Azure::Core::Test::TestMode::PLAYBACK: {
+          testModeValue = "PLAYBACK";
+        }
+        break;
+        case Azure::Core::Test::TestMode::LIVE:
+        default: {
+          testModeValue = "LIVE";
+        }
+        break;
+      };
+
+#if defined(_MSC_VER)
+      _putenv_s("AZURE_TEST_MODE", testModeValue.c_str());
```

Comment:
wooo, I would not expect a test to be setting env variables.  Why are you doing this?  

The env vars should be set by the user only.

Reply:
because on a system without those envs set the whole thing blows up, i need to ensure the env is set for these tests.

Reply:
and for these tests i need to control the test mode, this tests need to have a predictable behavior , relying on env for tests that are supposed to be predictable and dependable is a shoot yourself in the foot scenario, i do not want to give control of the behavior to an external entity, thus i need to control it.

Reply:
Let's bring it to the team.

As per the Testing with playback and recording design and strategy, these would be the requeriments:

- User can create new recordings at any given time for all tests by setting the test-env to RECORD
- Tests are expected to run on LIVE mode all the time as the preferred mode for validation.
- Tests should run on PLAYBACK mode only for running on public CI (cpp - keyvault - ci) or locally on devs machines,
- Tests should run LIVE mode on internal pipelines (cpp - keyvault). The internal pipeline is run before releasing a new version of the package. At that point, we can't rely on recordings. We want to run at least a LIVE test.

So, making tests to become hardcoded as PLAYBACK would potentially bring lack of security before releasing because we are not testing against the latest server responses.

Actually, there should be some strategy of updating all recordings periodically.

Reply:
yeah, let's bring it up in the standup on monday

Reply:
I would rather use the `tuesday` C++ sync meeting or schedule a new meeting with the team.  Rick's been trying so hard to make the standup quick